src/Lean/Server/Requests.lean

```diff
@@ -65,31 +63,42 @@ structure RequestContext where
   initParams    : Lsp.InitializeParams
 
 abbrev RequestTask α := Task (Except RequestError α)
+abbrev RequestT m := ReaderT RequestContext <| ExceptT RequestError m
```

Comment:
Is this all just for being able to call `readDoc` inside of `run*M` instead of outside?

Reply:
It's mostly so that we can type the argument to `runM` as `RequestT M`. Consequently the code inside can still access the request handler context, and throw request errors rather than only a generic `IO.Error`.

Reply:
Oh I see, it's about exceptions. As you may know I'd like to avoid duplicate exception layers, but I guess that's not realistic here. What error codes other than `internalError` are you expecting widget handlers to use?

Reply:
At least `invalidParams` for incorrect arguments.

Reply:
I had hoped that would be covered by the deserialization, but I guess the param type will not necessarily describe all argument constraints. I'm still not sure if it's worth the (not terribly big) code complication. The RPC interface of a widget should probably be considered private to the widget, so it's unlikely the client code will expect and handle invalidParams specially, no?

Reply:
I could definitely see people writing widget-support RPC libraries which can be used by multiple JS codebases, in which case some kind of error handling needs to happen. For example, our own [getWidgetSource](https://github.com/leanprover/lean4/pull/1407/files#diff-1a6eaee5d12ad418d2657e155b36b353081ee0bc8fefbe6628c6ea0779d1258eR108) is now a bit simpler since it can run in `CoreM` and use all the standard functions rather than accessing parts of the `Snapshot` data.